packages/block-editor/src/components/link-control/index.js

```diff
@@ -26,16 +26,21 @@ import { useSelect } from '@wordpress/data';
 import LinkControlSettingsDrawer from './settings-drawer';
 import LinkControlSearchItem from './search-item';
 import LinkControlSearchInput from './search-input';
+import LinkControlSearchCreate from './search-create-button';
 
 function LinkControl( {
 	value,
 	settings,
 	onChange = noop,
 	showInitialSuggestions,
+	showCreatePages,
+	createEmptyPage,
```

Comment:
I think we may need to make this more generalised. Something along the lines of `handleCreateEmpty`. This is because `LinkControl` won't always be creating Pages as that is very tied to the Nav Block use case.

@aduth Has mentioned we need to be sure not couple `LinkControl` to the Nav Block too tightly.

Reply:
I'm still thinking about this - are pages really connected **only** with the navigation block? I can easily imagine you might want to have similar functionality in other blocks containing links (a row of buttons comes to mind). 

What if we kept the entity creation code in `LinkControl` and only add a new property to it that will allow such functionality to show up? It will be disabled by default so usages of LinkControl would need to opt into the feature. It could be something in lines of `allowCreatingEntities` with a Boolean value (or maybe a list of supported post types like `[ 'page' ]`)

Reply:
@marekhrabe I've also been puzzling over the same thing.

@aduth What do you think about this?

Reply:
> I'm still thinking about this - are pages really connected **only** with the navigation block? I can easily imagine you might want to have similar functionality in other blocks containing links (a row of buttons comes to mind).

This may be true, but at least from the perspective of what the block editor is intended to represent (and support), there is no concept of a "page" that it can or should understand to exist.

This is reflected in the relationship between the primary modules making up the editor screen:

- [`block-editor`](https://github.com/WordPress/gutenberg/tree/master/packages/block-editor): A dumb block editor
- [`editor`](https://github.com/WordPress/gutenberg/tree/master/packages/editor): A block editor wired to read from and save to a post in WordPress
- [`edit-post`](https://github.com/WordPress/gutenberg/tree/master/packages/edit-post): The post editor screen

It feels to me that we probably don't want to be baking this into the LinkControl at all, but rather have some way of managing this sort of thing from where it is used (in the individual blocks, or perhaps some abstraction of `LinkControl` within the `editor` package which could be more closely associated to how it interacts with pages).

At its most generic, would it be something where we need to expose some options for manipulating (in this case adding) the suggestions which are presented to the user, and the specific behaviors associated when those suggestions are selected?

I see this was changed recently to something more generic "create entity". But what should that be interpreted to mean in the context of a dumb block editor? I suppose that there might be something which we could bake in here if we could generalize this behavior of "search term -> create new thing".

Reply:
> I suppose that there might be something which we could bake in here if we could generalize this behavior of "search term -> create new thing".

I feel like the only use case we're aware of that might want to create Entities on the fly is Navigation. Therefore if we make this an optional handler prop that you pass in then we're keeping `LinkControl` "dumb" whilst still affording flexibility.

That said, it is worth bearing in mind that the `fetchSearchSuggestions` has both a baked-in implementation and the ability to over as a prop. Perhaps we should take a similar approach with `Page` creation baked in via wrapping the component and passing in a handler, whilst also allowing for overides if the prop has already been provided.

Reply:
I think it's fairly reasonable that the component has this generalized understanding of searching for results associated with the input value, and that we would expect this to be optional and customized by the consumer (where in our case, the consumer is the `editor` package, which provides this as a setting, and handles it by searching for and providing posts).

So that's all fine to me.

And the more I think on it, I think a similar case can be made about presenting an option to create some new _thing_ in response to the search input, in a way which can likewise be generalized enough to not tie it to the specific posts use-case, but still support it.

---

That being said, one alternative that we could potentially consider is to try to fit this into the existing concept of a suggestion, and specifically the suggestions which the `editor` is already providing. In other words, construct the suggestion for a new post [to include alongside the search results](https://github.com/WordPress/gutenberg/blob/e18e0a0e8b8470fc77e61e673410a50f14a6238a/packages/editor/src/components/provider/index.js#L28-L45), then handle how choosing that suggestion should create a new post [when the value is changed](https://github.com/WordPress/gutenberg/blob/e18e0a0e8b8470fc77e61e673410a50f14a6238a/packages/block-library/src/navigation-link/edit.js#L183-L204). I think if we could conceptually agree that a suggestion could serve this purpose, but we're not totally happy on the idea of implementing this in the block (i.e. not reusable), then we could always consider some further abstraction of "a link control which also handles page creation" (in `editor`).

Reply:
> That being said, one alternative that we could potentially consider is to try to fit this into the existing concept of a suggestion, and specifically the suggestions which the editor is already providing. In other words, construct the suggestion for a new post to include alongside the search results, then handle how choosing that suggestion should create a new post when the value is changed.

This feels like it would be possible to move the "Create" suggestion into the Editor fetch results. After all I've already promoted the "Create" button to be part of the suggestions by manually appending it. 

However, I'm trying to see the benefit of this over simply [manually appending a create result to the API results as we currently do](https://github.com/WordPress/gutenberg/pull/19775/files#diff-106913d5079b57c6032c0775292d8d9dR121).

Ultimately you still have a valid "suggestion" which you end up treating differently when rendered. You also have a means to provide a specific handler to deal with the creation rather than have it baked in with the API results which could be slightly opaque.

The main difference appears to be that you're proposing handling the page creation in the `onChange` callback being passed into `LinkControl`. When `onChange` is called with a `type` of `__CREATE_TYPE__` then you'd presumably create the entity and update the `value` passed into `LInkControl`. What this doesn't cover is how we would control the UI state that needs to change as a result of this process being async rather than sync. For example, [when creating a page we need to show a loading state whilst the page is being created](https://github.com/WordPress/gutenberg/pull/19775/files#diff-106913d5079b57c6032c0775292d8d9dR183). Could we default `onChange` to being an `async` function and then always `await` it, setting loading state until a response is received?

One benefit of the approach you suggest would be keyboard handling of the `form` `onSubmit`. Currently, we expect the same handler we use for standard suggestions to work for "CREATE" suggestions. This doesn't work at all and so currently the submit with `[ENTER]` is completely broken.

I'd really value your input here. Much appreciated.

Reply:
I like the suggestion to make `onChange` async and resolve immediately for known values and await if we need to do more work (like creating a page). This sounds like it could be opened to "creating" more types of things in future and would definitely level all the different suggestions so there is no separate code path just for creating pages but it's all better integrated.

Reply:
I'm not sure I like the idea of an asynchronous `onChange`, because it seems to imply there's a guaranteed relationship between `onChange` and some future value. These sorts of assumptions can cause issues like the one currently being addressed in #19856. The component should be able to work fine without any `onChange` or with an `onChange` that does something other than set the next value, and it should handle a `value` which could have nothing to do with what `onChange` was provided.

I see the challenge here in that there will be some delay between when the user selects the suggestion and when the value can be updated to reflect the URL of the newly-created page.

For that, I wonder if we could do one of:

- Try to optimistically update the value to one which is "transient" (temporary), but is sufficient to represent the link in presenting it to the user as the selected item. We do something similar with image uploads in the editor, using [temporary blob URLs](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL) for the image. I'm not sure what options we might have for representing the not-yet-created page as a valid URL, and these sorts of temporary values can be troublesome if an autosave happens to occur while in its temporary state.
   - If we can settle on a generalized idea for what it means for a "value" to be transient, I think it would be okay to bake in some specialized logic into LinkControl that presents it in an appropriate manner.
- Allow some way to express that LinkControl is in a read-only or indeterminate (loading) state that can communicate that the page is being created (like a [`readOnly` prop](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly)). Perhaps combine this with something like #19856, where the input field will be shown as long as the `value` is still `undefined` (even if the user had "submitted" their selection).
- Expose `LinkControl`'s _form submission_ in a way that the default handling (calling `onChange` immediately) can be prevented and you can substitute your own, where presumably your own would be to set the new `value` only after the page has been created.

Touching back on the earlier idea of "presenting an option to create some new thing in response to the search input, in a way which can likewise be generalized enough to not tie it to the specific posts use-case, but still support it". If we did something like this, it could make sense that the link control would be aware that "creating some new thing" might need to be asynchronous, and it could then handle more of the presentational aspects of displaying a loading indicator. Something like an async `onCreate` wouldn't be as problematic as an async `onChange`, since it's not tied to an assumption about the next value. I expect the hypothetical `onCreate` would resolve to a new value (suggestion) and that could be what's passed through to `onChange`.

Reply:
@aduth Thanks your detailed input. Very much appreciated and great to get more perspetive.

Following this, my understanding of the best route forward that satisfies all the requirements I"m aware of (both technical and design) is:

* leave `onChange` as it currently is - accepting a suggestion value and being fully synchronous. 
* add a new prop `onCreate` specifically designed to handle creating new items via the `LinkControl`.
* `onCreate` would be _explicitly_ async and could do anything it liked so long as it resolved with a valid `suggestion`.
* Internally we'd have a method such as `handleOnCreate` which would `await` the `onCreate` prop's value and then (if valid) call `onChange` with that value.
* `LinkControl`'s internal UI state would be handled by `handleOnCreate` so as to show appropriate loading indicators whilst any async creation occurs ([as required by the Design](https://github.com/WordPress/gutenberg/issues/18900#issuecomment-578200190)).

Hopefully this will also solve the thorny issue of needing to ensure that form submissions via the keyboard on "Create" suggestions can be handled in the same way as normal suggestions.

Reply:
@getdave Yes, that seems like a reasonable path forward to me.

Reply:
I've tweaked the existing implementation to match [my suggestion](https://github.com/WordPress/gutenberg/pull/19775#discussion_r371178624). We can discuss in a follow-up review.